### PR TITLE
Port back D47532

### PR DESCRIPTION
**Pull request by martong**

### CODE REVIEW

File: unittests/AST/ASTImporterTest.cpp [hunk 1]
```diff
@@ -2278,72 +2274,74 @@ TEST_P(ImportFunctions, ImportDefinitions) {
...
 
-TEST_P(ImportFunctions,
-       ImportPrototypeThenDefinition) {
+TEST_P(ImportFunctions, ImportPrototypeThenDefinition) {
   auto Pattern = functionDecl(hasName("f"));
 
   {
     Decl *FromTU = getTuDecl("void f();", Lang_CXX, "input0.cc");
-    auto FromD = FirstDeclMatcher<FunctionDecl>().match(FromTU, Pattern);
+    FunctionDecl *FromD =
```
Comment (balazske) [inline unittests/AST/ASTImporterTest.cpp]: Why is this not `auto*`?

File: lib/AST/DeclBase.cpp [hunk 2]
```diff
@@ -1371,7 +1375,7 @@ static bool shouldBeHidden(NamedDecl *D) {
   // from being visible?
   if (isa<ClassTemplateSpecializationDecl>(D))
     return true;
-  if (FunctionDecl *FD = dyn_cast<FunctionDecl>(D))
+  if (auto *FD = dyn_cast<FunctionDecl>(D))
```
Comment (balazske) [inline lib/AST/DeclBase.cpp:1378]: Is this change intentional?

Comment (martong) [reply lib/AST/DeclBase.cpp:1378]: Yes, it is. We should avoid repetition of the type which is already there in the `dyn_cast`'s template params.

File: unittests/AST/ASTImporterTest.cpp (hunk 1 above)
Comment (martong) [reply unittests/AST/ASTImporterTest.cpp]: Good catch, is should be auto*.

File: lib/AST/DeclBase.cpp (hunk 2 above)
Comment (martong) [reply lib/AST/DeclBase.cpp:1378]: The upstream has `auto *` as well.

**Review by balazske — APPROVED** [review]
